### ISSUE TITLE
CEM explainer Slow with Pytorch

### ISSUE THREAD

**Issue by EoinKenny** [SeldonIO/alibi#217]
Hi, thanks a lot for the library, I'm sure it was a tonne or work and it looks great.

I've been using the CEM explainer (and Prototype one) for MNIST, and I've noticed that when I follow the documentation tutorial it's quite fast < 10sec

But when I insert a black box model (in my case a Pytorch CNN object) which isn't Keras or Tensorflow, it's incredibly slower, like sometimes up to 5-6 hours to get an explanation, or as little as 5-6mins also.

I am 100% sure it's coded correctly, my black-box Pytorch function takes the same input as the Keras model, and outputs then same probabilities in numpy in the correct shape etc.

It just seems that the library is optimized for Keras/TF, is that correct? Any way around this or have I made an error do you think?

Thank you for any advice/help.

**Comment by arnaudvl** [comment]
Hi, thanks for the kind words. When you pass a Keras/TensorFlow model, the whole optimization problem is defined within the TensorFlow graph, making automatic differentiation possible. When you pass a PyTorch model though, the graph sees it as a black box and needs to estimate numerical gradients. This requires a lot of calls to the model and is slow. You can also find this in the docs: https://docs.seldon.io/projects/alibi/en/latest/methods/CFProto.html#Numerical-Gradients. If you use the counterfactual, you could choose to remove the prediction loss term by setting ```c_init``` to 0 and ```c_steps``` to 1, but this could deteriorate the quality of the outcome. Another option is to increase the value for ```update_num_grad``` to *N* to only evaluate the numerical gradients every *N* optimization steps. Yet another option is to change ```eps``` to a larger value to obtain potentially bigger gradients and finding a counterfactual quicker.

In the future, we'd like to have an alternative for PyTorch models so they are also treated as "white-box" and/or use genetic optimization to avoid the numerical gradient computations for black box models.

**Comment by jklaise** [comment]
Just to elaborate on the plans, we do want to support both Tensorflow and Pytorch models for these explanation methods in the future but we can't provide any estimates of when just yet. There is a project to first port the TF1.x implementations to TF2.x (https://github.com/SeldonIO/alibi/projects/3), after this it should be relatively easy to do a Pytorch implementation as well since TF2.x and Pytorch syntax is very similar.

That being said we're very open to contributions if anyone fancies a challenge!
